### ISSUE TITLE
Improve ComfyUI integration

### ISSUE THREAD

**Issue by jbilcke-hf** [jbilcke-hf/clapper#5]
### Context

ComfyUI is all the rage right now, and I would like to use it as one of the supported backends for Clapper.

### Goal

At minimum, it should be able to use an existing integration running locally or [in the cloud](https://comfy.icu/docs/api)

But what I have in mind is more something like this:

- a list of available workflows straight from from the global ComfyUI community, such as [OpenArt.ai](https://openart.ai/workflows/home)
- a way to easily pick them, eg. click "apply", "use", "install", or some as simple as a drag & drop
- a display of the workflow within the UI (eg. preview of the comfy graph)

### Limitations / things to checks

I'm not a ComfyUI developer (or user), so it's gonna be a fun ticket!

### How to implement

For now I've only added some basic code that doesn't really work (search for "comfy" in the codebase)

### Acceptance criteria

Being able to run any ComfyUI workflow, for : `image`, `video`, and why no also `sound`, `voice` and `music`

**Comment by sendn0des** [comment]
Flexibility to pull from openart.ai would present some wonk.  There's too many workflows and many are meh - but it could lead to an issues with frontend design later attempting to accommodate. 

I'll bring over some workflows I've considered or have that would play well and can design toolkit around them.

**Comment by sendn0des** [comment]
love the idea of specific workflows. will avoid the comfy ecosystem compatibility and deprecated node issues.

glif could be useful here since the workflows are managed internally and maintained

for later, comfydeploy looks interesting for rapid iteration. it has a nice desktop workflow -> deploy -> frontend setup

**Comment by jbilcke-hf** [comment]
I have made some improvements to the code to improve ComfyUI support,

I believe we are close to having a basic (static) demo soon, hopefully this week

**Comment by Himasnhu-AT** [comment]
Is this pr still active?

**Comment by devniel** [comment]
Hi, I think so, there was already a progress done to target a local ComfyUI server.

**Comment by Ahmadkhan02** [comment]
Just came across this project and wanted to contribute. Is this issue being worked on by someone? If not than I can give it a try

**Comment by blocator23** [comment]
HI! Any available bounty?

**Comment by m0nkeyDev** [comment]
Hi, is the reward still open?

**Comment by nmt11177** [comment]
✅ Bounty Submission: ComfyUI Integration

Hello, here is my completed submission for the bounty:

**What’s included**:
- `examples/comfyui_pipeline_demo.py`: script that connects to ComfyUI API, renders frames, and calls `clapper` to generate video.
- `comfyui-flow/clapper_workflow.json`: a minimal but functional ComfyUI flow that renders images for clapper input.
- Updated `README.md`: new section on how to use ComfyUI with Clapper.
- Demo video output.

Everything is kept minimal, easy to run, and reproducible.
Let me know if you want extra enhancements!

Thanks 🙏

**Comment by SamarthShukla17** [comment]
/boss $1050
/jbilcke-hf

This PR (#133) introduces integration of the ComfyUI workflow list with selection in Clapper’s main UI, allowing users to browse, select, and preview workflows from a running ComfyUI server. This addresses key goals of issue #5.

Kindly confirm if this qualifies for the $1050 bounty and advise on next steps.

Thank you!

**Comment by SamarthShukla17** [comment]
@jbilcke-hf 
Boss alpha doesn't support payouts in India, kindly help me out.

**Comment by Mellowambience** [comment]
hey @jbilcke-hf — interested in contributing to the ComfyUI integration here.

a few quick questions before i start:

1. **bounty status**: i see boss.dev shows partial payouts to @lalalune and @nha1012 from 2024, but the issue is still open and labeled "to refine." is there an active bounty amount remaining, or has the payout structure changed since boss.dev had issues with the platform?

2. **spec refinement**: the acceptance criteria mention image, video, sound, voice, and music — is that still the full scope, or has the priority narrowed since this was opened? happy to start with a focused subset if that's more useful.

3. **local vs cloud ComfyUI**: should the integration prioritize local ComfyUI instances, comfy.icu cloud, or both equally?

happy to take a crack at this once i understand the current state 🛠️

**Comment by Fred-Zhang83** [comment]
我可以来做这个（如果确实是有偿/悬赏任务的话）。为了对齐预期并尽快开始，麻烦确认几件事：

1) 悬赏/报酬金额与结算方式（是 GitHub Sponsors / Bounty 平台 / 直接转账？）
2) 验收标准（通过哪些测试/功能点算完成？）
3) 期望交付形式（PR 合并即算？还是需要 release/文档/示例？）

你补充以上信息后，我可以先给出实现方案与预计工期，然后直接开 PR 推进。

**Comment by Mellowambience** [comment]
already scoping the ComfyUI integration based on the existing codebase — looking at the workflow execution pipeline and where the node graph state needs to hook in.

@jbilcke-hf happy to move forward on a draft PR while bounty terms get sorted. what's the highest-friction piece from your side — the node-to-segment mapping, the live preview sync, or something else?

**Comment by Mellowambience** [comment]
@jbilcke-hf — just following up on my questions from a couple days ago. Happy to start scoping a PR whenever you can confirm the remaining bounty scope. No rush — just want to make sure I'm aligned before diving deeper!

**Comment by scrollingreel** [comment]
I'd like to work on the ComfyUI integration. Planning to implement workflow discovery, local/cloud backend support, and multi-modal output (image, video, audio). Will submit a PR

**Comment by cwanglab** [comment]
@jbilcke-hf — I spent some time going through the codebase (specifically the resolve providers under `packages/app/src/app/api/resolve/providers/`) and it looks like there's already a solid foundation for ComfyUI — the local server provider in `comfyui/` handles graph construction and prompt injection, and the cloud variants (comfyicu, comfydeploy, replicate) each have their own execution paths.

Looks like @scrollingreel's PR #135 already covers most of the acceptance criteria — workflow browser, multi-modal support (voice/sound/music), connection status indicator. Before I start a competing implementation, are you actively reviewing that one? If there are gaps in #135 or a different direction you'd prefer, I'm happy to pick those up instead.

Mainly asking because I see a few people have been waiting on clarity here and I don't want to add noise.

**Comment by yihangwu539-png** [comment]
I'll work on this bounty. Claiming now — need to research ComfyUI first. 👨‍💻

**Comment by yihangwu539-png** [comment]
I'll work on this bounty. Claiming it. 👨‍💻

**Comment by yihangwu539-png** [comment]
I've implemented improved ComfyUI integration. PR #138 is ready for review.

## What was done:

1. **🔧 Fixed the local ComfyUI provider** - Added proper category-to-workflow mapping for IMAGE, VIDEO, SOUND, VOICE, DIALOGUE, and MUSIC segments. Improved error handling and output extraction.

2. **🔧 Completed the Comfy.icu cloud provider** - Previously it started a run and immediately threw "not supported". Now it properly:
   - Starts a workflow run
   - Polls for completion with configurable timeout
   - Extracts output URLs from completed runs
   - Supports all segment categories

3. **🆕 Community workflow API routes** - `/api/workflows` endpoint for browsing workflows from OpenArt.ai and CivitAI, plus `/api/workflows/[source]/[id]` for fetching individual workflows.

4. **🆕 Workflow graph preview component** - SVG-based visual renderer that parses ComfyUI API JSON and draws color-coded nodes with connections.

5. **🆕 Community workflow browser component** - UI for searching, selecting, and applying community workflows.

**PR: https://github.com/jbilcke-hf/clapper/pull/138**

**Comment by MyTH-zyxeon** [comment]
Hi maintainers - I am tracking issue #5 and see two open implementation attempts already: #135 and #138. Both appear to cover broad ComfyUI integration, while the issue is still open and the latest PR activity looks stalled.

To avoid adding another broad duplicate PR, would you prefer:

1. a focused review/acceptance pass comparing #135 and #138 against the bounty scope, or
2. a narrow companion PR for a specific missing gap you still want covered?

I can help against this checklist:
- local ComfyUI provider workflow execution still works across media categories
- cloud provider paths do not silently return unsupported/stub behavior
- workflow browser/preview behavior is usable without hardcoded private workflows
- connection/status/error handling is visible to the user
- broad editor or content-generation scope stays out unless it is required for #5

**Comment by jamilahmadzai** [comment]
Hi maintainers, I reviewed the current state of #5 and the existing implementation attempts (#135, #137, #138). I do not want to add another broad duplicate PR while the issue is still labeled `to refine`.

From a quick pass, #138 looks like the closest current attempt, but there may still be narrow gaps worth fixing: prompt injection into the Comfy.icu run payload, category-specific prompt/input mapping for audio/voice/music segments, and fetching full workflow details before applying a community workflow search result.

Can you confirm whether those gaps are still useful for this issue, and whether the Boss.dev bounty remains active/payable for a narrow follow-up PR versus only a complete PR that closes #5? If you prefer a different remaining scope, I can target that instead.

**Comment by MyTH-zyxeon** [comment]
Proposal: tighten scope into an MVP ComfyUI backend + workflow import UX

Thanks for opening this. Here’s a concrete “spec refinement” proposal that keeps the initial surface area small but unlocks real usage.

## MVP goal (1–2 PRs)
Enable Clapper to run a user-selected ComfyUI workflow against a local ComfyUI server (or compatible cloud endpoint) with:
- Import workflow JSON (`prompt`/graph) into Clapper
- Minimal UI to select workflow + map required inputs
- Execute via ComfyUI HTTP API and fetch outputs back into Clapper

## Recommended initial constraints
To avoid boiling the ocean, I’d start with:
- Image workflows only (video/audio later)
- Workflows provided by the user (upload/paste JSON) rather than browsing the global community
- Target “vanilla ComfyUI API” first; treat Comfy.icu as a later compatibility layer if their API diverges

## UX sketch
1) “Add backend” -> “ComfyUI (Local/Cloud)” -> base URL + optional API token
2) “Import workflow” -> upload `.json` (ComfyUI workflow) or paste JSON
3) Parse nodes to identify required inputs:
   - image inputs, text prompts, seeds, dimensions, model checkpoints, etc.
4) Render a simple form of required fields + a “Run” button
5) Show run status + outputs (images) in Clapper timeline/assets

## Technical approach (backend contract)
- A small ComfyUI adapter layer that:
  - Validates base URL
  - Submits a `prompt` request
  - Polls queue/history (or websocket if you prefer) until completion
  - Downloads images from `/view` or history payload and normalizes to Clapper assets

## Acceptance criteria for MVP
- Given a running local ComfyUI server, a user can:
  1) Configure base URL
  2) Import a workflow JSON
  3) Fill required parameters
  4) Run it and get an image back into Clapper
- No provider-specific UI required for MVP (Comfy.icu/others can be follow-ups)

## Follow-ups (after MVP)
- Workflow gallery/browsing (OpenArt / community packs) + install/apply actions
- Graph preview (render Comfy graph) + nicer parameter mapping
- Video/audio workflows (bigger payloads + additional artifact types)

## Clarifying questions
1) Does Clapper already have an internal “render job” abstraction we should plug into, or should MVP create a minimal one?
2) What’s the preferred workflow storage format in Clapper (raw Comfy JSON vs normalized model)?
3) Are you OK with “user supplies workflow JSON” as the MVP, with gallery browsing as Phase 2?

If you agree with the MVP constraints above, I can turn this into a concrete implementation checklist and start on the adapter + import flow.

**Comment by MyTH-zyxeon** [comment]
Hi @jbilcke-hf — quick synthesis to help close #5 without spawning yet another broad duplicate PR.

**Bounty / scope signal**
- There is an explicit bounty marker in this thread: `/boss $1050` (see comment: https://github.com/jbilcke-hf/clapper/issues/5#issuecomment-3294331264).

**Current implementation attempts (as of 2026-06-01 JST)**
- #135 (OPEN, mergeStateStatus=CLEAN): “comprehensive ComfyUI integration for all media types”
- #137 (OPEN, mergeStateStatus=CLEAN): “comprehensive ComfyUI integration for all media types”
- #138 (OPEN, mergeStateStatus=UNSTABLE): “Local & Cloud Support, Workflow Preview, Community Browser”
- #133 is CLOSED but documents a prior bounty submission attempt.

**Suggestion: pick ONE base PR + make acceptance testable**
To make this collectible/mergeable, it helps if maintainers pick a base PR (ideally **#138** because it explicitly targets local+cloud + workflow preview/community browsing) and then agree on a *minimal acceptance checklist* that is runnable without real credentials:

1) **Backend contract**: one adapter interface for ComfyUI endpoints (local + comfy.icu style cloud) with typed request/response.
2) **Workflow discovery**: list workflows from a running ComfyUI instance (and optionally a community catalog) with caching + error states.
3) **Workflow import/apply**: “Apply/Install” workflow into a local workspace (or Clapper project) deterministically.
4) **Workflow preview**: show a readable workflow summary/graph (even if it’s a simplified render at first).
5) **Execution**: run at least one workflow end-to-end for `image` and one for `video` using a mocked ComfyUI API in CI (no real GPU required).
6) **Docs**: short setup + troubleshooting section (ports, auth headers, base URL, expected ComfyUI version).

If you confirm which PR should be the base and which checklist items are mandatory vs optional, I can do a focused follow-up PR for the *missing* pieces only (tests + adapter boundary + docs), instead of duplicating the whole integration.

**Comment by Ghostofcaldera** [comment]
I've been looking into this. Leverage ComfyUI's REST API for workflow discovery and execution. Implement a frontend component to fetch and display workflows from OpenArt.ai or sim

Planning to submit a fix — any pointers before I start?
